### PR TITLE
Remove in-tree support for Networking and rely on extension controllers

### LINKED ISSUE #1109: Provide Network Extensibility
As a user operating a Gardener landscape I want to have a choice of using the optimal Kubernetes network provider for the used cloud provider(s) as well as the best network provider for my particular use case.

We currently only support Calico as network provider. This is for example problematic for Azure where it is not officially supported (although it does work). New and specialised network providers are emerging (https://kubernetes.io/docs/concepts/cluster-administration/networking/ now lists 23 providers) and we should enable them in a flexible way.

There is a proposal at https://github.com/gardener/gardener/blob/master/docs/proposals/03-networking.md

### PR DESCRIPTION
**Pull request by zanetworker**
**What this PR does / why we need it**:
This PR integrate the networking extension into Gardener.

**Which issue(s) this PR fixes**:
Fixes #1109 

**Special notes for your reviewer**: 
:warning: For local development you need to update your provider extension charts and use at least version `0.10.0-dev-0cf466fc06c6a185a6bdfb143fc82bcd377ed234`. We'll create a new release of the gardener-extensions repository before releasing gardener.

N.B: this is dependant on another PR on the extensions repo: https://github.com/gardener/gardener-extensions/pull/162 

**Release note**:
```action operator
:warning: Gardener does no longer have in-tree supported for managing the shoot's network plugin. Instead, it now relies on extension controllers to deploy the network plugin for the specific provider they have been developed for, e.g. Calico, Flannel, etc. (see [GEP-3 on network extensibility](https://github.com/gardener/gardener/blob/master/docs/proposals/03-networking.md)). We have prepared the implementation of the [Calico networking extension](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/networking-calico). After updating Gardener you need to create a corresponding `ControllerRegistration` resource to make this extension controller known to Gardener (otherwise none of your shoots will be able to get reconciled anymore). Please find example `ControllerRegistration` resources [here (Calico)](https://github.com/gardener/gardener-extensions/blob/0.10.0/controllers/networking-calico/example/controller-registration.yaml). To get information about more details please walk through [these documents](https://github.com/gardener/gardener/tree/master/docs/extensions).
```

### CODE REVIEW

File: pkg/controllermanager/controller/shoot/shoot_control_delete.go [hunk 1]
```diff
@@ -210,6 +210,16 @@ func (c *Controller) runDeleteShootFlow(o *operation.Operation) *gardencorev1alp
 			Fn:           flow.TaskFn(botanist.DeleteClusterAutoscaler).RetryUntilTimeout(defaultInterval, defaultTimeout),
 			Dependencies: flow.NewTaskIDs(initializeShootClients),
 		})
+		destroyNetwork = g.Add(flow.Task{
+			Name:         "Destroying shoot networking",
+			Fn:           flow.TaskFn(botanist.DestroyNetwork).RetryUntilTimeout(defaultInterval, defaultTimeout),
+			Dependencies: flow.NewTaskIDs(initializeShootClients),
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_control_delete.go]:
```suggestion
			Dependencies: flow.NewTaskIDs(cleanKubernetesResources),
```

We may only destroy the networking after the cluster has been successfully cleaned.

File: pkg/controllermanager/controller/shoot/shoot_control_delete.go [hunk 2]
```diff
@@ -210,6 +210,16 @@ func (c *Controller) runDeleteShootFlow(o *operation.Operation) *gardencorev1alp
...
+		})
+		_ = g.Add(flow.Task{
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_control_delete.go]: Can you add this task to `syncPointCleaned`?

File: pkg/controllermanager/controller/shoot/shoot_control_reconcile.go [hunk 3]
```diff
@@ -216,6 +216,16 @@ func (c *Controller) runReconcileShootFlow(o *operation.Operation, operationType
 			Fn:           flow.TaskFn(botanist.ComputeShootOperatingSystemConfig).RetryUntilTimeout(defaultInterval, defaultTimeout),
 			Dependencies: flow.NewTaskIDs(initializeShootClients, waitUntilInfrastructureReady),
 		})
+		deployNetworking = g.Add(flow.Task{
+			Name:         "Deploying Cluster Network Resources",
+			Fn:           flow.TaskFn(botanist.DeployNetwork).RetryUntilTimeout(defaultInterval, defaultTimeout).SkipIf(o.Shoot.HibernationEnabled),
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_control_reconcile.go]:
```suggestion
			Fn:           flow.TaskFn(botanist.DeployNetwork).RetryUntilTimeout(defaultInterval, defaultTimeout),
```

File: pkg/controllermanager/controller/shoot/shoot_maintenance_control.go [hunk 4]
```diff
@@ -195,7 +195,7 @@ func (c *defaultMaintenanceControl) Maintain(shootObj *gardenv1beta1.Shoot, key
 
 		delete(s.Annotations, common.ShootOperation)
 
-		controllerutils.AddTasks(s.Annotations, common.ShootTaskDeployInfrastructure, common.ShootTaskDeployKube2IAMResource)
+		controllerutils.AddTasks(s.Annotations, common.ShootTaskDeployInfrastructure, common.ShootTaskDeployNetwork, common.ShootTaskDeployKube2IAMResource)
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_maintenance_control.go]: Why is this added here? Remove it please.

File: pkg/operation/shoot/types.go [hunk 5]
```diff
@@ -44,6 +44,7 @@ type Shoot struct {
 
 	CloudConfigMap       map[string]CloudConfig
 	Extensions           map[string]Extension
+	NetworkStatus        []byte
```
Comment (rfranzke) [inline pkg/operation/shoot/types.go]: What do we use this for? Let's remove it, it's not needed right now.

File: pkg/operation/common/types.go [hunk 6]
```diff
@@ -221,12 +221,6 @@ const (
 	// KubeSchedulerServerName is the name of the kube-scheduler server.
 	KubeSchedulerServerName = "kube-scheduler-server"
 
-	// CalicoKubeControllersDeploymentName is the name of calico-kube-controllers deployment.
```
Comment (rfranzke) [inline pkg/operation/common/types.go]: Let's keep these deployment/daemonset names. We need them for the health check. Yes, it's not perfect because we cannot assume that calico is deployed and if any other extension is used the health check will always fail. However, we don't have a mechanism yet to allow extensions to report health status. Also, there is no other network extension yet. So, let's keep it for now until we can work on this problem (only after the API migration topic).

File: pkg/operation/common/types.go [hunk 7]
```diff
@@ -337,6 +328,12 @@ const (
 	// ShootTaskDeployInfrastructure is a name for a Shoot's infrastructure deployment task.
 	ShootTaskDeployInfrastructure = "deployInfrastructure"
 
+	// ShootTaskDeployNetwork is a name for a Shoot's Network deployment task.
+	ShootTaskDeployNetwork = "deployNetwork"
```
Comment (rfranzke) [inline pkg/operation/common/types.go]: Not needed, remove it.

File: pkg/operation/common/types.go [hunk 8]
```diff
@@ -337,6 +328,12 @@ const (
...
+	// ShootNetworkTypeCalico is used to identify the calico network type
+	// TODO: remove once the new shoot API is introduced, value will be fetched from spec.networking.type
+	ShootNetworkTypeCalico = "calico"
```
Comment (rfranzke) [inline pkg/operation/common/types.go]: I'd not create a constant for this, let's rather 'hard-code' it where it's used.

File: pkg/operation/botanist/controlplane.go [hunk 9]
```diff
@@ -68,6 +68,8 @@ func (b *Botanist) DeployNamespace(ctx context.Context) error {
 			gardencorev1alpha1.BackupProvider: string(b.Seed.CloudProvider),
 			gardencorev1alpha1.SeedProvider:   string(b.Seed.CloudProvider),
 			gardencorev1alpha1.ShootProvider:  string(b.Shoot.CloudProvider),
+			// TODO: remove once the new shoot API is introduced, value should be fetched from spec.networking.type
```
Comment (rfranzke) [inline pkg/operation/botanist/controlplane.go]: Why? I think it still makes sense to keep labels on the namespaces for the various provider types. Probably we should also add another label of the operating system config types, etc. It's useful for webhooks.

File: pkg/operation/botanist/botanist_suite_test.go [hunk 10]
```diff
@@ -243,26 +243,20 @@ var _ = Describe("health check", func() {
 		}
 
 		// system component deployments
-		calicoKubeControllersDeployment = newDeployment(shootNamespace, common.CalicoKubeControllersDeploymentName, common.GardenRoleSystemComponent, true)
```
Comment (rfranzke) [inline pkg/operation/botanist/botanist_suite_test.go]: Let's keep the calico health checks for now, see reasoning below.

File: pkg/operation/botanist/network.go [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+// for a successful reconciliation of a network resource.
+const NetworkDefaultTimeout = 10 * time.Minute
+
+// DeployNetwork creates the `Network` extension resource in the shoot namespace in the seed
+// cluster. Gardener waits until an external controller did reconcile the cluster successfully.
+func (b *Botanist) DeployNetwork(ctx context.Context) error {
+	var (
+		lastOperation                = b.Shoot.Info.Status.LastOperation
+		creationPhase                = lastOperation != nil && lastOperation.Type == gardencorev1alpha1.LastOperationTypeCreate
+		requestNetworkReconciliation = creationPhase || controllerutils.HasTask(b.Shoot.Info.Annotations, common.ShootTaskDeployNetwork)
```
Comment (rfranzke) [inline pkg/operation/botanist/network.go]: This is not needed. I suggest "copying" the deploy function from the worker resource, for example.

File: pkg/operation/botanist/network.go [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+			metav1.SetMetaDataAnnotation(&network.ObjectMeta, gardencorev1alpha1.GardenerOperation, gardencorev1alpha1.GardenerOperationReconcile)
+		}
+		return nil
+	})
+}
+
+// DestroyNetwork deletes the `Network` extension resource in the shoot namespace in the seed cluster,
+// and it waits for a maximum of 10m until it is deleted.
+func (b *Botanist) DestroyNetwork(ctx context.Context) error {
+	if err := b.K8sSeedClient.Client().Delete(ctx, &extensionsv1alpha1.Network{ObjectMeta: metav1.ObjectMeta{Namespace: b.Shoot.SeedNamespace, Name: b.Shoot.Info.Name}}); err != nil && !apierrors.IsNotFound(err) {
```
Comment (rfranzke) [inline pkg/operation/botanist/network.go]:
```suggestion
	return client.IgnoreNotFound(b.K8sSeedClient.Client().Delete(ctx, &extensionsv1alpha1.Network{ObjectMeta: metav1.ObjectMeta{Namespace: b.Shoot.SeedNamespace, Name: b.Shoot.Info.Name}}))
```

File: pkg/operation/botanist/network.go [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+		network := &extensionsv1alpha1.Network{}
+		if err := b.K8sSeedClient.Client().Get(ctx, client.ObjectKey{Name: b.Shoot.Info.Name, Namespace: b.Shoot.SeedNamespace}, network); err != nil {
+			return retry.SevereError(err)
+		}
+
+		if err := health.CheckExtensionObject(network); err != nil {
+			b.Logger.WithError(err).Error("Network did not get ready yet")
+			return retry.MinorError(err)
+		}
+		if providerStatus := network.Status.ProviderStatus; providerStatus != nil {
```
Comment (rfranzke) [inline pkg/operation/botanist/network.go]: Not needed to store this, we don't use it anywhere.

File: pkg/operation/botanist/network.go [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+		network = &extensionsv1alpha1.Network{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      b.Shoot.Info.Name,
+				Namespace: b.Shoot.SeedNamespace,
+			},
+			Spec: extensionsv1alpha1.NetworkSpec{
+				PodCIDR:     string(b.Shoot.GetPodNetwork()),
+				ServiceCIDR: string(b.Shoot.GetServiceNetwork()),
+			},
```
Comment (rfranzke) [inline pkg/operation/botanist/network.go:53]: Why not added the `networking` section right now to the shoot API and default the `type` with `calico` if it's not set? That makes things easier, I guess.

**Review by rfranzke — CHANGES_REQUESTED** [review]

**Comment by rfranzke** [comment]
Also, can you please add documentation to the `docs/extensions` folder?

File: pkg/operation/botanist/controlplane.go (hunk 9 above)
Comment (zanetworker) [reply pkg/operation/botanist/controlplane.go]: I did not mean to remove the label from the namespace. I meant to pick up the value of the label from the shoot `spec.networking.type` section. Now it is Calico, but different shoots can choose different networking plugins right?

Comment (rfranzke) [reply pkg/operation/botanist/controlplane.go]: Then please enhance the comment, I didn't get what was meant. Also, consider introducing `spec.networking` already today, this will ease the implementation at various places, I guess.

File: pkg/apis/garden/types.go [hunk 15]
```diff
@@ -619,6 +619,8 @@ type ShootSpec struct {
 	Hibernation *Hibernation
 	// Kubernetes contains the version and configuration settings of the control plane components.
 	Kubernetes Kubernetes
+	// Networking contains information about cluster networking such as CNI Plugin type, CIDRs, ...etc.
+	Networking Networking
```
Comment (rfranzke) [inline pkg/apis/garden/types.go]:
```suggestion
	Networking *Networking
```

File: pkg/apis/garden/types.go [hunk 16]
```diff
@@ -657,6 +659,19 @@ type ShootStatus struct {
...
+
+const (
+	Calico NetworkType = "calico"
+)
+
+// Networking defines networking parameters for the shoot cluster.
+type Networking struct {
+	gardencore.K8SNetworks
+	// Type identifies the type of the networking plugin
+	Type NetworkType
```
Comment (rfranzke) [inline pkg/apis/garden/types.go]: We should add provider config here (optional).

File: pkg/apis/garden/v1beta1/defaults.go [hunk 17]
```diff
@@ -45,8 +48,14 @@ func SetDefaults_Shoot(obj *Shoot) {
 		if cloud.AWS.Networks.Nodes == nil {
 			if cloud.AWS.Networks.VPC.CIDR != nil {
 				obj.Spec.Cloud.AWS.Networks.Nodes = cloud.AWS.Networks.VPC.CIDR
+				if networking.Nodes == nil {
+					obj.Spec.Networking.Nodes =  cloud.AWS.Networks.VPC.CIDR
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/defaults.go]: +1, nice. format pls.

File: pkg/apis/garden/v1beta1/types.go [hunk 18]
```diff
@@ -704,6 +705,20 @@ type ShootStatus struct {
...
+
+const (
+	Calico NetworkType = "calico"
+)
+
+// Networking defines networking parameters for the shoot cluster.
+type Networking struct {
+	gardencorev1alpha1.K8SNetworks  `json:",inline"`
+	// Type identifies the type of the networking plugin
+	Type NetworkType `json:"type"`
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/types.go]: provider config is missing as well

File: pkg/apis/garden/v1beta1/types.go [hunk 19]
```diff
@@ -658,10 +658,11 @@ type ShootSpec struct {
 	Hibernation *Hibernation `json:"hibernation,omitempty"`
 	// Kubernetes contains the version and configuration settings of the control plane components.
 	Kubernetes Kubernetes `json:"kubernetes"`
+	// Networking contains information about cluster networking such as CNI Plugin type, CIDRs, ...etc.
+	Networking Networking `json:"networking"`
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/types.go]:
```suggestion
	// +optional
	Networking *Networking `json:"networking,omitempty"`
```

File: pkg/apis/garden/v1beta1/defaults.go [hunk 20]
```diff
@@ -114,6 +142,16 @@ func SetDefaults_Shoot(obj *Shoot) {
 		}
 	}
 
+	if networking.Pods == nil {
+		obj.Spec.Networking.Pods = &defaultPodCIDR
+	}
+	if networking.Services == nil {
+		obj.Spec.Networking.Services = &defaultServiceCIDR
+	}
+	if len(networking.Type) == 0 {
```
Comment (rfranzke) [inline pkg/apis/garden/v1beta1/defaults.go]: Let's also default the `nodes` field of `.spec.networking`.

File: pkg/controllermanager/controller/shoot/shoot_control_reconcile.go [hunk 21]
```diff
@@ -216,6 +216,16 @@ func (c *Controller) runReconcileShootFlow(o *operation.Operation, operationType
 			Fn:           flow.TaskFn(botanist.ComputeShootOperatingSystemConfig).RetryUntilTimeout(defaultInterval, defaultTimeout),
 			Dependencies: flow.NewTaskIDs(initializeShootClients, waitUntilInfrastructureReady),
 		})
+		deployNetworking = g.Add(flow.Task{
+			Name:         "Deploying Cluster Network Resources",
+			Fn:           flow.TaskFn(botanist.DeployNetwork).RetryUntilTimeout(defaultInterval, defaultTimeout),
+			Dependencies: flow.NewTaskIDs(initializeShootClients, waitUntilInfrastructureReady, computeShootOSConfig),
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_control_reconcile.go]:
```suggestion
			Dependencies: flow.NewTaskIDs(deployGardenerResourceManager),
```

Should be sufficient, or? Why should we wait for the OS config or infrastructure?

**Review by rfranzke — CHANGES_REQUESTED** [review]

File: pkg/controllermanager/controller/shoot/shoot_control_reconcile.go (hunk 21 above)
Comment (zanetworker) [reply pkg/controllermanager/controller/shoot/shoot_control_reconcile.go]: lets also wait for infrastructure and OSConfig. With `usePodCIDR` for IPAM the value is picked from the `Node.podCIDR` and I guess nodes will not join without proper infra and OS config.

**Review by rfranzke — APPROVED** [review]
